backend/plugin/db/hive/hive.go

```diff
@@ -0,0 +1,181 @@
+package hive
+
+import (
+	"context"
+	"database/sql"
+	"io"
+	"strconv"
+	"time"
+
+	"github.com/beltran/gohive"
+	"github.com/pkg/errors"
+
+	"github.com/bytebase/bytebase/backend/plugin/db"
+
+	storepb "github.com/bytebase/bytebase/proto/generated-go/store"
+	v1pb "github.com/bytebase/bytebase/proto/generated-go/v1"
+)
+
+type Driver struct {
+	config   db.ConnectionConfig
+	ctx      db.ConnectionContext
+	dbClient *gohive.Connection
+}
+
+var (
+	_ db.Driver = (*Driver)(nil)
+)
+
+func (d *Driver) Open(_ context.Context, _ storepb.Engine, config db.ConnectionConfig) (db.Driver, error) {
+	// field legality check.
+	if config.Username == "" {
+		return nil, errors.Errorf("user not set")
+	}
+	if config.Host == "" {
+		return nil, errors.Errorf("hostname not set")
+	}
+	d.config = config
+	d.ctx = config.ConnectionContext
+
+	// initialize database connection.
+	configuration := gohive.NewConnectConfiguration()
+	port, err := strconv.Atoi(config.Port)
+	if err != nil {
+		return nil, errors.Errorf("conversion failure for 'port' [string -> int]")
+	}
+	// TODO(tommy): actually there are various kinds of authentication to choose among [SASL, KERBEROS, NOSASL, PLAIN SASL]
+	// "NONE" refers to PLAIN SASL that doesn't need authentication.
+	authMethods := "NONE"
+	conn, errConn := gohive.Connect(config.Host, port, authMethods, configuration)
+	if errConn != nil {
+		return nil, errors.Errorf("failed to establish connection")
+	}
+	d.dbClient = conn
+	return d, nil
+}
+
+func (d *Driver) Close(_ context.Context) error {
+	err := d.dbClient.Close()
+	if err != nil {
+		return errors.Errorf("faild to close connection")
+	}
+	return nil
+}
+
+func (d *Driver) Ping(ctx context.Context) error {
+	if d.dbClient == nil {
+		return errors.Errorf("database not connected")
+	}
+	if _, err := d.Execute(ctx, "SELECT 1", db.ExecuteOptions{}); err != nil {
+		return errors.Errorf("bad connection")
+	}
+	return nil
+}
+
+func (*Driver) GetType() storepb.Engine {
+	return storepb.Engine_HIVE
+}
+
+func (*Driver) GetDB() *sql.DB {
+	return nil
+}
+
+func (d *Driver) Execute(ctx context.Context, statement string, _ db.ExecuteOptions) (int64, error) {
+	var rowCount int64
+	cursor := d.dbClient.Cursor()
```

Comment:
defer cursor.Close()

Reply:
thx, that's an elegant way to release resource...